rewriter/GenCallAsm.cpp

```diff
@@ -280,21 +280,309 @@ static std::string sig_string(const CAbiSignature &sig,
   return ss.str();
 }
 
+static void call_string(
+    const std::optional<std::string> target_name,
+    WrapperKind kind,
+    AsmWriter &aw,
+    Arch arch) {
+
+  add_comment_line(aw, "Call wrapped function");
+
+  if (arch == Arch::X86) {
+    if (!target_name) {
+      // indirect call
+      assert(kind == WrapperKind::IndirectCallsite);
+      add_asm_line(aw, "call *%r12");
+    } else {
+      // direct call
+      add_asm_line(aw, "call "s + target_name.value());
+    }
+  } else if (arch == Arch::Aarch64) {
+    if (!target_name) {
+      // indirect call
+      assert(kind == WrapperKind::IndirectCallsite);
+      add_asm_line(aw, "br "s + target_name.value());
+    } else {
+      // direct call
+      add_asm_line(aw, "b "s + target_name.value());
+    }
+  }
+}
+
+static AsmWriter get_asmwriter(bool as_macro) {
+  std::string terminator = {};
+  // Code generated as a macro needs to terminate each line with '\'
+  if (as_macro) {
+    terminator = "\\"s;
+  }
+  return {.ss = {}, .terminator = terminator};
+}
+
+static void switch_stacks(AsmWriter &aw, uint32_t caller_pkey, uint32_t target_pkey, Arch arch) {
+  if (arch == Arch::X86) {
+    // Save the old frame pointer and set the frame pointer for the call gate
+    add_asm_line(aw, "pushq %rbp");
+    add_asm_line(aw, "movq %rsp, %rbp");
+    // Save registers that are preserved across function calls before switching to
+    // the other compartment's stack. This is on the caller's stack so it's not in
+    // the diagram above.
+    for (auto &r : preserved_registers) {
+      add_asm_line(aw, "pushq %"s + r);
+    }
+
+    uint32_t caller_pkru = ~((0b11 << (2 * caller_pkey)) | 0b11);
+    add_raw_line(aw, llvm::formatv("ASSERT_PKRU({0:x8}) \"\\n\"", caller_pkru));
+    // Change pkru to the intermediate value before copying args
+    add_comment_line(aw, "Set PKRU to the intermediate value to move arguments");
+    // wrpkru requires zeroing rcx and rdx, but they may have arguments so use r10
+    // and r11 as scratch registers
+    add_asm_line(aw, "movq %rcx, %r10");
+    add_asm_line(aw, "movq %rdx, %r11");
+    emit_mixed_wrpkru(aw, caller_pkey, target_pkey);
+    add_asm_line(aw, "movq %r10, %rcx");
+    add_asm_line(aw, "movq %r11, %rdx");
+
+    // Switch stacks to the target stack
+    emit_switch_stacks(aw, caller_pkey, target_pkey, "r11"s);
+  } else if (arch == Arch::Aarch64) {
+    // no-op for now
```

Comment:
nit: I added warnings while working on the ARM stuff for things that we want to punt for now. I think `#warning` doesn't make sense in this case, but you could log using `llvm::errs() << $MSG` to make sure we don't forget anything. Just a nit since this is a WIP anyway.

Reply:
I threw in llvm errrs for all the ARM no-ops.